### ISSUE TITLE
Fix broken links to .md files

### ISSUE THREAD

**Issue by MarkLodato** [slsa-framework/slsa#938]
There are quite a few broken links in our site, most of which to non-existent `.md` files. The problem is that we have been using default [jekyll-relative-links](https://github.com/benbalter/jekyll-relative-links) plugin, which automatically converts a link to `file.md` to `file` **if and only if `file.md` exists**. If the author uses an absolute path rather than relative (e.g. [`/spec/v1.0/levels.md`](https://github.com/slsa-framework/slsa/blob/main/docs/how-to-orgs.md?plain=1#L13) rather than `spec/v1.0/levels.md`) or if the file gets moved and the link does not get updated, then Jekyll will silently link to the non-existent `.md` file.

I suggest a two-part fix:

1. Stop using jekyll-relative-links because it leads to too much confusion. Fix up all existing cases and disable the plugin.
2. Set up [netlify-plugin-checklinks](https://github.com/munter/netlify-plugin-checklinks#readme) or similar to check links for us, so that we don't have so many broken links going forward.

**Comment by MarkLodato** [comment]
A quick way to find broken `.md` links is to search through the built HTML files.

```bash
bundle exec jekyll build
git grep --no-index 'href="[^h][^"]*\.md' _site
```

**Comment by joshuagl** [comment]
I've sent a PR to fix .md links in #946 which addresses the first part of the suggested fix.

Initial experimentation with [netlify-plugin-checklinks](https://github.com/munter/netlify-plugin-checklinks#readme) has been error-prone, but I will dig into it more as time allows.

**Comment by MarkLodato** [comment]
I was wrong about using an absolute link to a `.md` file - it actually works. Example: https://slsa.dev/how-to-orgs

Source:

```md
For all [SLSA levels](/spec/v1.0/levels.md), you follow the same steps:
```

Result:

```html
<p>For all <a href="/spec/v1.0/levels">SLSA levels</a>, you follow the same steps:</p>
```

That said, there are still many remaining broken links (posted on #946) and the source code to use the URL instead of the `.md` file seems like a good idea regardless.

**Comment by MarkLodato** [comment]
> I've sent a PR to fix .md links in #946 which addresses the first part of the suggested fix.

Note that that PR only changes a few links to `.md` files, whereas we have many:

```bash
$ git grep -e '^\[.*\]: [^:#]*\.md$' --or -e '\]([^:#)]*\.md[#)]' docs | wc -l
172
```

> Initial experimentation with [netlify-plugin-checklinks](https://github.com/munter/netlify-plugin-checklinks#readme) has been error-prone, but I will dig into it more as time allows.

An alternative (non-netlify-specific) tool might also work, which we could run in GitHub Actions.

**Comment by MarkLodato** [comment]
Looks like the thinks that are broken due to this specific issue are limited:

```bash
$ git grep --color --no-index 'href="[^h][^"]*\.md' _site
_site/blog.html:the valuable feedback we received on the <a href="2023-02-24-slsa-v1-rc.md">first release candidate</a>. This is
_site/blog.html:      <p>Interested in getting involved? Now’s the chance to <a href="2023-02-24-slsa-v1-rc.md">provide your feedback on the foundational v1 release of the SLSA framework.</a></p>
_site/example.html:<p>SLSA 4 <a href="requirements.md">requires</a> two-party source control and hermetic builds.
_site/how-to-orgs.html:<a href="/provenance/v1.md">https://slsa.dev/provenance/</a>.</p>
_site/spec/v1.0-rc2/onepage.html:<p>SLSA v1.0 does not address this threat, but it may be addressed in a <a href="#future-directions.md">future
_site/spec/v1.0-rc2/onepage.html:<p>SLSA v1.0 does not address this threat, but it may be addressed in a <a href="#future-directions.md">future
_site/spec/v1.0-rc2/threats.html:<p>SLSA v1.0 does not address this threat, but it may be addressed in a <a href="future-directions.md">future
_site/spec/v1.0-rc2/threats.html:<p>SLSA v1.0 does not address this threat, but it may be addressed in a <a href="future-directions.md">future
_site/spec/v1.0/onepage.html:<p>SLSA v1.0 does not address this threat, but it may be addressed in a <a href="#future-directions.md">future
_site/spec/v1.0/onepage.html:<p>SLSA v1.0 does not address this threat, but it may be addressed in a <a href="#future-directions.md">future
_site/spec/v1.0/threats.html:<p>SLSA v1.0 does not address this threat, but it may be addressed in a <a href="future-directions.md">future
_site/spec/v1.0/threats.html:<p>SLSA v1.0 does not address this threat, but it may be addressed in a <a href="future-directions.md">future
```

Specifically:

- example.md: `requirements.md` no longer exists (fixed in #946)
- threats.md (v1.0 and v1.0-rc2): `future-directions.md` doesn't work for some unknown reason (fixed in #946)
- how-to-orgs.md: `/provenance/v1.md` no longer exists due to #939
- autogenerated pages (blog.html and onepage.html) don't work with the plugin, so links to .md files are not converted there

**Comment by joshuagl** [comment]
>     * autogenerated pages (blog.html and onepage.html) don't work with the plugin, so links to .md files are not converted there

these are resolved by the changes in #946 

```
% git grep --color --no-index 'href="[^h][^"]*\.md' _site
_site/how-to-orgs.html:<a href="/provenance/v1.md">https://slsa.dev/provenance/</a>.</p>
```

That how-to-orgs page doesn't appear to be resolved by #939, but it's unclear why the link to an .md file 7 lines earlier is converted?!

**Comment by MarkLodato** [comment]
> > ```
> > * autogenerated pages (blog.html and onepage.html) don't work with the plugin, so links to .md files are not converted there
> > ```
> 
> these are resolved by the changes in #946

Ah, I'm sorry. I didn't understand that from the PR. Could you explain in the PR description?


> That how-to-orgs page doesn't appear to be resolved by #939, but it's unclear why the link to an .md file 7 lines earlier is converted?!

Yeah, it's a mystery why sometimes the conversion works and sometimes it doesn't!

**Comment by joshuagl** [comment]
I've updated #946 to only address the broken links in the generated site. For this issue I plan to

- [ ] remove all uses of .md file extensions in relative links
- [ ] remove jekyll-relative-links from our site's configuration
- [ ] implement a broken link checker for our generated site, to be run in CI

**Comment by marcelamelara** [comment]
@lehors Given the recent PR, I worry this issue may be causing more confusion. I suspect you checked all the links with your website generation update?

**Comment by lehors** [comment]
@marcelamelara I agree but I must admit not to have had this issue on my radar. I will take another look and see if any of it is still relevant and what we might do about it if needed.
I just realized that I checked that the generated website is the same as the old one but so that means any broken link we had is still broken. :-/

**Comment by lehors** [comment]
So, upon further investigation I can confirm that the netlify link checker plugin isn't reliable. It actually crashed when I tried it! And judging by the number of open issues against it I don't think that's going to be the way to go.
We should investigate using another tool that we could run in GitHub Actions as suggested by Mark.

For now, I have fixed all our internal broken links so I think we can close this issue and open another one to track setting up an automatic link checker if we want to.

**Comment by lehors** [comment]
@SunShineHead Go away. Remove this repository (https://github.com/slsa-framework/slsa) from your tracking list. Do not post any more comments, issues, or PRs to this repository.
